keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+complete, and lacks support recursive data types (OpenAPIv3 & CRD structural schemas are not
+possible to validate with CEL).
+
+### Risks and Mitigations
+
+#### Accidental misuse
+
+Break the control plane by consuming excessive CPU and/or memory the api-server.
+
+Mitigation: CEL is specifically designed to constrain the running time of expressions and to limit
+the memory utilization. Since CRD creation is a privileged operation, it should be safe to
+integrate.
+
+#### Malicious use
+
+Breaking out of the sandbox to run untrusted code in the apiserver or exfiltrate data.
+
+Mitigation: CEL is designed to sandbox code execution. 
+
+
+### Future Plan
+  
+#### Defaulting
+
+The `x-kubernetes-default` extension will be used. The location of the defaulter
+determines the scope of defaulter.
+
+The 'field' specifies which field is to be defaulted using a field path that is
+relative to the scope root. The defaulter will only be run if 'field' is unset, 
+and the result of the expression must match the type of the field to be
+defaulted.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          x-kubernetes-default:
+		    - rule: "!has(spec.type) && spec.Type == 'LoadBalancer' ? 'ServiceExternalTrafficPolicy' : null"
+			  field: spec.externalTrafficPolicy
+```
+
+- Each defaulter may have multiple rules.
+
+- The 'field' is optional and defaults to the scope if not explicitly set.
+
+If the expression evaluates to null, the field is left unset.
+TODO: Any downsides to using null this way?
+
+#### Conversion
+
+Conversion rules will be made available via a ConversionRules strategy. A set of
+rules can be provided for each supported source/destination version pair. Each rule will
+specify which field it sets using a field path. A 'scope' field *pattern* may
+also be used specify all locations in the object that the conversion rule should
+be applied.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+ conversion:
+    strategy: ConversionRules
+	converters:
+		- fromVersion: v1beta1
+		  toVersion: v1
+		  rules:
+		    - field: spec.xyz.podIPs
+			  rule: "has(spec.xyz.podIP) ? [spec.xyz.podIP] : []"
+		- fromVersion: v1
+		  toVersion: v1beta1
+		  rules:
+		    - field: xyz.podIP
+		      scope: spec
+			  rule: "size(xyz.podIPs) > 0 ? xyz.podIPs[0] : null"
+```
+
+Expressions that return object types will overwrite all fields that they return and
+leave all other fields to be auto-converted.
+
+The 'scope' is optional and defaults to the object root.
+
+The 'field' is optional and defaults to the scope? TODO: is it better to make it required for conversion?
+
+If the expression evaluates to null, the field is left unset.
+
+TODO: demonstrate writing fields into annotations for round trip
+
+TODO: demonstrate string <-> structured (label selector example & ServicePort example)
+
```

Comment:
demonstrate renaming a field (would have to set a new field `bar` and unset the old field `foo`)

Reply:
I've trimmed the CRD conversion section and dropped all the implementation details so we can sort it out properly as a separate KEP. Agree that renaming needs to be really convenient.